src/psyclone/domain/common/algorithm/psyir.py

```diff
@@ -36,18 +36,80 @@
 '''This module contains PSyclone Algorithm-layer-specific PSyIR classes.
 
 '''
-from psyclone.psyir.nodes import Call, Reference, DataNode
-from psyclone.psyir.symbols import TypeSymbol
+from psyclone.psyir.nodes import Call, Reference, DataNode, Literal, \
+    ArrayReference
+from psyclone.psyir.symbols import TypeSymbol, ContainerSymbol, \
+    GlobalInterface, RoutineSymbol
 from psyclone.errors import GenerationError
 
 
 class AlgorithmInvokeCall(Call):
-    '''An invoke call in a PSyclone Algorithm layer.'''
+    '''An invoke call in a PSyclone Algorithm layer.
 
+    :param invoke_routine_symbol: the routine that this call calls.
+    :type invoke_routine_symbol: \
+        py:class:`psyclone.psyir.symbols.RoutineSymbol`
+    :param int index: the position of this invoke call relative to \
+        other invokes in the algorithm layer.
+    :param parent: optional parent of this node in the PSyIR. Defaults \
+        to None.
+    :type parent: sub-class of :py:class:`psyclone.psyir.nodes.Node` \
+        or NoneType
+
+    :raises TypeError: if the index argument is not an integer.
+    :raises ValueError: if the index argument is negative.
+
+    '''
     _children_valid_format = "[KernelFunctor]*"
     _text_name = "AlgorithmInvokeCall"
     _colour = "green"
 
+    def __init__(self, invoke_routine_symbol, index, parent=None):
+        super(AlgorithmInvokeCall, self).__init__(
+            invoke_routine_symbol, parent=parent)
+
+        if not isinstance(index, int):
+            raise TypeError(
+                "AlgorithmInvokeCall index argument should be an int but "
+                "found '{0}'.".format(type(index).__name__))
+        if index < 0:
+            raise ValueError(
+                "AlgorithmInvokeCall index argument should be a non-negative "
+                "integer but found {0}.".format(index))
+
+        self._index = index
+        self._language_level_routine_symbol = None
```

Comment:
I was going to argue with you about this but actually you are completely right - the symbols are PSy-layer symbols. The only reason we are storing them in the algorithm layer is to avoid clashes. However, languages usually allow renaming when importing, so we could avoid clashes that way instead. Therefore, a more logical long term solution would be to generate the symbols in the PSy-layer and pass them to the algorithm lower method. Anyway, given that renaming is not something we support at the moment, I will plough ahead with the current approach regardless, and rename as you suggest.